README.md
# Lexicoid for Ruby

Have you ever needed to convert a timestamp into something short, memorable, easy to look at, and easy to type? Look no further! You want a "lexicoid."

Inspired by [Brandur Leach](https://brandur.org/fragments/base32-slugs), Lexicoid for Ruby gives you lexicographically sortable, time-based identifiers that are easy to read and easy to type; perfect for using in URLs and other places that take human input.

Use lexicoids for anywhere that you have a timestamp that operates on "human scale" … think microblog status identifiers or URL-encoded timestamps. Avoid them for distributed systems; instead, consider something like [KSUID](https://github.com/michaelherold/ksuid-ruby) for that use case.

## Installation

Install the gem and add it to the application's Gemfile by executing:

    $ bundle add lexicoid

If you are not using Bundler to manage dependencies, install the gem by executing:

    $ gem install lexicoid

## Usage

Generate a lexical identifier for the current moment:

```ruby
Lexicoid.now
```

Convert any time-related object to a lexical identifier:

```ruby
# Integer
Lexicoid.from(1_678_767_177)

# Float
Lexicoid.from(1_678_767_285.293477)

# Time
Lexicoid.from(Time.now)

# DateTime
Lexicoid.from(DateTime.now)

# ActiveSupport::TimeWithZone
Lexicoid.from(2.weeks.ago)
```

There is also a `Kernel` method available for converting objects to lexicoids:

```ruby
Lexicoid(Time.now)

Lexicoid("oops") #=> raises ArgumentError

Lexicoid("oops", exception: false) #=> ""
```

## Contributing

So you're interested in contributing to Lexicoid? Check out our [contributing guidelines](CONTRIBUTING.md) for more information on how to do that.

## Supported Ruby Versions

This library aims to support and is [tested against](https://github.com/michaelherold/lexicoid-ruby/actions) the following Ruby versions:

* Ruby 3.0
* Ruby 3.1
* Ruby 3.2

If something doesn't work on one of these versions, it's a bug.

This library may inadvertently work (or seem to work) on other Ruby versions, however we will only provide support for the versions listed above.

If you would like this library to support another Ruby version or implementation, you may volunteer to be a maintainer. Being a maintainer entails making sure all tests run and pass on that implementation. When something breaks on your implementation, you will be responsible for providing patches in a timely fashion. If critical issues for a particular implementation exist at the time of a major release, we may drop support for that Ruby version.

## Versioning

This library aims to adhere to [Semantic Versioning 2.0.0](http://semver.org/spec/v2.0.0.html). Report violations of this scheme should as bugs. Specifically, if a minor or patch version breaks backward compatibility, that version should be immediately yanked and/or a new version should be immediately released that restores compatibility. Only new major versions will introduce breaking changes to the public API. As a result of this policy, you can (and should) specify a dependency on this gem using the [pessimistic version constraint](http://guides.rubygems.org/patterns/#pessimistic-version-constraint) with two digits of precision. For example:

    spec.add_dependency "lexicoid", "~> 0.1"

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

We expect everyone interacting in the Lexicoid for Ruby project's codebases, issue trackers, chat rooms and mailing lists to follow the [code of conduct](https://github.com/michaelherold/lexicoid/blob/main/CODE_OF_CONDUCT.md).

## Acknowledgments

[Brandur Leach](https://brandur.org/fragments/base32-slugs) wrote about the concept behind this gem and inspired its creation.

[Luciano Mammino](https://loige.co/) took the first step, created a [Rust crate](https://crates.io/crates/lexicoid) and came up with the [excellent name](https://github.com/lmammino/lexicoid) for these identifiers.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0](https://github.com/michaelherold/lexicoid-ruby/tree/0.1.0) - 2023-03-19

### Added

- Converting timestamps in the form of `Integer`s, `Float`s, `Time`s, `DateTime`s, and anything responding to `#to_date`, including `ActiveSupport::TimeWithZone` via `Lexicoid.from`.
- Generating the lexicoid for the current moment with `Lexicoid.now`.
- Converting timestamps with the `Kernel#Lexicoid` method, with optional exception squelching via `Lexicoid(Time.now, exception: false)`.


test/test_helper.rb
# frozen_string_literal: true

if ENV["CI"] || ENV["COVERAGE"]
  require "simplecov"

  SimpleCov.start do
    add_filter "/test/"
  end

  SimpleCov.command_name "test:unit"
end

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)

require "minitest/autorun"
require "minitest/reporters"
require "pry"
require "pry-byebug"
require "time"

require "active_support"
require "active_support/core_ext/integer/time"
require "active_support/core_ext/time"
require "active_support/time_with_zone"

require "lexicoid"

Time.zone = "UTC"

# Remove this when bumping to ActiveSupport 7.1
ActiveSupport::TimeWithZone.singleton_class.remove_method(:name)

Minitest::Reporters.use! [
  Minitest::Reporters::DefaultReporter.new(
    color: true
  )
]


test/test_lexicoid.rb
# frozen_string_literal: true

require "test_helper"

class TestLexicoid < Minitest::Test
  def test_that_it_has_a_version_number
    refute_nil ::Lexicoid::VERSION
  end

  def test_from_with_integers
    assert_equal "22", Lexicoid.from(0)               # Thu Jan 01 1970 00:00:00 -0000
    assert_equal "gk", Lexicoid.from(100)             # Thu Jan 01 1970 00:01:40 -0000
    assert_equal "6wc2", Lexicoid.from(10000)         # Thu Jan 01 1970 02:46:40 -0000
    assert_equal "2ykm2", Lexicoid.from(500000)       # Tue Jan 06 1970 18:53:20 -0000
    assert_equal "5bse2", Lexicoid.from(1700000)      # Tue Jan 20 1970 16:13:20 -0000
    assert_equal "2apny22", Lexicoid.from(28000000)   # Sat Nov 21 1970 01:46:40 -0000
    assert_equal "6567f22", Lexicoid.from(550000000)  # Sat Jun 06 1987 17:46:40 -0000
    assert_equal "flllz22", Lexicoid.from(1550000000) # Tue Feb 12 2019 19:33:20 -0000
    assert_equal "gehebv2", Lexicoid.from(1654301676) # Sat Jun 04 2022 00:14:36 -0000
    assert_equal "gei4p52", Lexicoid.from(1654401676) # Sun Jun 05 2022 04:01:16 -0000
    assert_equal "gj7x3v2", Lexicoid.from(1674301676) # Sat Jan 21 2023 11:47:56 -0000
    assert_equal "gj7x3vc", Lexicoid.from(1674301677) # Sat Jan 21 2023 11:47:57 -0000

    assert_equal "26222222", Lexicoid.from(2**32)          # Sun Feb 07 2106 06:28:16 -0000
    assert_equal "jzzzzzzzzzzzy", Lexicoid.from(2**63 - 1) # Sun Dec 04 292277026596 15:30:07 -0000
  end

  def test_from_with_floats
    assert_equal "22", Lexicoid.from(0.0)               # Thu Jan 01 1970 00:00:00 -0000
    assert_equal "gk", Lexicoid.from(100.0)             # Thu Jan 01 1970 00:01:40 -0000
    assert_equal "6wc2", Lexicoid.from(10000.0)         # Thu Jan 01 1970 02:46:40 -0000
    assert_equal "2ykm2", Lexicoid.from(500000.0)       # Tue Jan 06 1970 18:53:20 -0000
    assert_equal "5bse2", Lexicoid.from(1700000.0)      # Tue Jan 20 1970 16:13:20 -0000
    assert_equal "2apny22", Lexicoid.from(28000000.0)   # Sat Nov 21 1970 01:46:40 -0000
    assert_equal "6567f22", Lexicoid.from(550000000.0)  # Sat Jun 06 1987 17:46:40 -0000
    assert_equal "flllz22", Lexicoid.from(1550000000.0) # Tue Feb 12 2019 19:33:20 -0000
    assert_equal "gehebv2", Lexicoid.from(1654301676.0) # Sat Jun 04 2022 00:14:36 -0000
    assert_equal "gei4p52", Lexicoid.from(1654401676.0) # Sun Jun 05 2022 04:01:16 -0000
    assert_equal "gj7x3v2", Lexicoid.from(1674301676.0) # Sat Jan 21 2023 11:47:56 -0000
    assert_equal "gj7x3vc", Lexicoid.from(1674301677.0) # Sat Jan 21 2023 11:47:57 -0000
  end

  def test_from_with_datetimes
    assert_equal "22", Lexicoid.from(DateTime.parse("Thu Jan 01 1970 00:00:00 -0000"))
    assert_equal "gk", Lexicoid.from(DateTime.parse("Thu Jan 01 1970 00:01:40 -0000"))
    assert_equal "6wc2", Lexicoid.from(DateTime.parse("Thu Jan 01 1970 02:46:40 -0000"))
    assert_equal "2ykm2", Lexicoid.from(DateTime.parse("Tue Jan 06 1970 18:53:20 -0000"))
    assert_equal "5bse2", Lexicoid.from(DateTime.parse("Tue Jan 20 1970 16:13:20 -0000"))
    assert_equal "2apny22", Lexicoid.from(DateTime.parse("Sat Nov 21 1970 01:46:40 -0000"))
    assert_equal "6567f22", Lexicoid.from(DateTime.parse("Sat Jun 06 1987 17:46:40 -0000"))
    assert_equal "flllz22", Lexicoid.from(DateTime.parse("Tue Feb 12 2019 19:33:20 -0000"))
    assert_equal "gehebv2", Lexicoid.from(DateTime.parse("Sat Jun 04 2022 00:14:36 -0000"))
    assert_equal "gei4p52", Lexicoid.from(DateTime.parse("Sun Jun 05 2022 04:01:16 -0000"))
    assert_equal "gj7x3v2", Lexicoid.from(DateTime.parse("Sat Jan 21 2023 11:47:56 -0000"))
    assert_equal "gj7x3vc", Lexicoid.from(DateTime.parse("Sat Jan 21 2023 11:47:57 -0000"))
  end

  def test_from_with_times
    assert_equal "22", Lexicoid.from(Time.parse("Thu Jan 01 1970 00:00:00 -0000"))
    assert_equal "gk", Lexicoid.from(Time.parse("Thu Jan 01 1970 00:01:40 -0000"))
    assert_equal "6wc2", Lexicoid.from(Time.parse("Thu Jan 01 1970 02:46:40 -0000"))
    assert_equal "2ykm2", Lexicoid.from(Time.parse("Tue Jan 06 1970 18:53:20 -0000"))
    assert_equal "5bse2", Lexicoid.from(Time.parse("Tue Jan 20 1970 16:13:20 -0000"))
    assert_equal "2apny22", Lexicoid.from(Time.parse("Sat Nov 21 1970 01:46:40 -0000"))
    assert_equal "6567f22", Lexicoid.from(Time.parse("Sat Jun 06 1987 17:46:40 -0000"))
    assert_equal "flllz22", Lexicoid.from(Time.parse("Tue Feb 12 2019 19:33:20 -0000"))
    assert_equal "gehebv2", Lexicoid.from(Time.parse("Sat Jun 04 2022 00:14:36 -0000"))
    assert_equal "gei4p52", Lexicoid.from(Time.parse("Sun Jun 05 2022 04:01:16 -0000"))
    assert_equal "gj7x3v2", Lexicoid.from(Time.parse("Sat Jan 21 2023 11:47:56 -0000"))
    assert_equal "gj7x3vc", Lexicoid.from(Time.parse("Sat Jan 21 2023 11:47:57 -0000"))
  end

  def test_from_with_times_with_zone
    assert_equal "22", Lexicoid.from(Time.zone.parse("Thu Jan 01 1970 00:00:00 -0000"))
    assert_equal "gk", Lexicoid.from(Time.zone.parse("Thu Jan 01 1970 00:01:40 -0000"))
    assert_equal "6wc2", Lexicoid.from(Time.zone.parse("Thu Jan 01 1970 02:46:40 -0000"))
    assert_equal "2ykm2", Lexicoid.from(Time.zone.parse("Tue Jan 06 1970 18:53:20 -0000"))
    assert_equal "5bse2", Lexicoid.from(Time.zone.parse("Tue Jan 20 1970 16:13:20 -0000"))
    assert_equal "2apny22", Lexicoid.from(Time.zone.parse("Sat Nov 21 1970 01:46:40 -0000"))
    assert_equal "6567f22", Lexicoid.from(Time.zone.parse("Sat Jun 06 1987 17:46:40 -0000"))
    assert_equal "flllz22", Lexicoid.from(Time.zone.parse("Tue Feb 12 2019 19:33:20 -0000"))
    assert_equal "gehebv2", Lexicoid.from(Time.zone.parse("Sat Jun 04 2022 00:14:36 -0000"))
    assert_equal "gei4p52", Lexicoid.from(Time.zone.parse("Sun Jun 05 2022 04:01:16 -0000"))
    assert_equal "gj7x3v2", Lexicoid.from(Time.zone.parse("Sat Jan 21 2023 11:47:56 -0000"))
    assert_equal "gj7x3vc", Lexicoid.from(Time.zone.parse("Sat Jan 21 2023 11:47:57 -0000"))
  end

  def test_from_raises_for_improper_arguments
    ex = assert_raises(ArgumentError) { Lexicoid.from(Time.at(-1)) }
    assert_match(/after/, ex.message)

    ex = assert_raises(ArgumentError) { Lexicoid.from(DateTime.parse("Wed Dec 31 1969 23:59:59 -0000")) }
    assert_match(/after/, ex.message)

    ex = assert_raises(ArgumentError) { Lexicoid.from(1.day.before(Time.at(0))) }
    assert_match(/after/, ex.message)

    ex = assert_raises(ArgumentError) { Lexicoid.from(-1) }
    assert_match(/non-negative/, ex.message)

    ex = assert_raises(ArgumentError) { Lexicoid.from(-1.0) }
    assert_match(/non-negative/, ex.message)

    ex = assert_raises(ArgumentError) { Lexicoid.from("1234") }
    assert_match(/one of/, ex.message)
  end

  def test_kernel_method_delegates_correctly
    assert_equal Lexicoid(1234), Lexicoid.from(1234)
    assert_raises(ArgumentError) { Lexicoid.from(-1) }
  end

  def test_now_uses_the_current_time
    Time.stub :now, Time.at(1_678_943_067) do
      assert_equal "gkdeaqs", Lexicoid.now
    end
  end
end


test/core_ext/test_kernel.rb
# frozen_string_literal: true

require "test_helper"

class TestKernel < Minitest::Test
  def test_kernel_method_can_squelch_errors
    assert_raises(ArgumentError) { Lexicoid("oops") }
    assert_equal "", Lexicoid("oops", exception: false)
  end

  def test_kernel_method_is_private_on_objects
    assert_raises(NoMethodError) { Object.new.Lexicoid(Time.now) }
  end
end


lib/core_ext/kernel.rb
# frozen_string_literal: true

# The `Kernel` module is included by class `Object`, so its methods are available in every Ruby object.
#
# @see https://ruby-doc.org/current/Kernel.html
module Kernel
  private

  # Returns a lexicographically sortable friendly ID for a timestamp
  #
  # @api public
  # @since 0.1.0
  # @!visibility public
  #
  # @example Converts a post's created-at timestamp to a friendly ID
  #   Post = Struct.new(:created_at, keyword_init: true)
  #   post = Post.new(created_at: Time.parse("2023-03-19 14:33:01.80251 -0500"))
  #
  #   Lexicoid(post.created_at) #=> "gkfqavc"
  #
  # @param object [ActiveSupport::TimeWithZone, DateTime, Float, Integer, Time, #to_time]
  #   the object to convert into a lexicographically sorting friendly ID
  # @param exception [Boolean] raises an ArgumentError when the argument is
  #   improper and this is true, otherwise an improper argument leads to an empty string
  # @return [String] the friendly ID
  # @raise [ArgumentError] when the object is not a timestamp-like one
  def Lexicoid(object, exception: true)
    Lexicoid.from(object)
  rescue ArgumentError
    raise if exception
    ""
  end
end


lib/lexicoid.rb
# frozen_string_literal: true

require_relative "lexicoid/version"

# Generates lexicographically sortable, friendly IDs from timestamps
#
# A "lexicoid" is a short identifier derived from a timestamp for use in single
# source streaming systems. They are based on an [approach for giving friendly
# URLs to microblog posts, by Brandur Leach][1].
#
# In order to make them easy to use, lexicoids have the following properties:
#
# 1. They are lexicographically, i.e. alphabetically, sortable by time to the
#    second.
# 2. They encode to lower case letters for easier typing into a URL.
# 3. Their alphabet eliminates easily-confused letter combinations like `1`,
#    `l`, and `I`, or `0` and `O`.
#
# At their heart, lexicoids are a special case of [Base 32 encoding][2] for
# integers.
#
# Their properties makes them ideally suitable to human-scale use cases, like
# microblog post slugs or URL-encoded timestamps. They are poorly suited for
# distributed systems with multiple clocks or systems that generate items at a
# rate of more than one per second, regardless of the number of clocks.
#
# [1]: https://brandur.org/fragments/base32-slugs
# [2]: https://datatracker.ietf.org/doc/html/rfc4648
module Lexicoid
  # The Time instance representing the time at the beginning of the Unix epoch
  #
  # @api private
  # @private
  EPOCH = Time.at(0).utc.freeze

  # The alphabet for encoding lexicographic base32 IDs
  #
  # @api private
  # @private
  LEXICOGRAPHIC_BASE32 = "234567abcdefghijklmnopqrstuvwxyz"

  # Converts a timestamp into a lexicographically sortable, friendly ID
  #
  # @api public
  # @since 0.1.0
  #
  # @example Generate a lexicoid from a DateTime
  #   Lexicoid.from DateTime.parse("2023-03-19T21:19:50-05:00")
  #
  # @example Generate a lexicoid from a Float timestamp
  #   Lexicoid.from 2_147_483_647.0
  #
  # @example Generate a lexicoid from an Integer timestamp
  #   Lexicoid.from 946_684_799
  #
  # @example Generate a lexicoid from a Time
  #   Lexicoid.from Time.parse("2023-03-19 15:06:14.48414 -0500")
  #
  # @example Generate a lexicoid in a Rails application
  #   Lexicoid.from 2.weeks.ago
  #
  # @param number_or_time [ActiveSupport::TimeWithZone, DateTime, Float, Integer, Time, #to_time]
  #   the object to convert into a lexicographically sorting friendly ID
  # @return [String] the friendly ID
  # @raise [ArgumentError] when the object is unsuitable for conversion
  def self.from(number_or_time)
    number = maybe_coerce_time(number_or_time)
    bytes = bytes_from_number(number)
    buffer = encode(bytes)

    buffer.join
  end

  # Generates a lexicographically sortable, friendly ID from the current time
  #
  # @api public
  # @since 0.1.0
  #
  # @example Generate a friendly ID for the current moment
  #   Lexicoid.now
  #
  # @return [String] the friendly ID
  def self.now
    from Time.now
  end

  # Extracts the bytes from a number in big endian order
  #
  # @api private
  # @private
  #
  # @param number [Integer] the number to extract bytes from
  # @return [Array<Integer>] the bytes from the number as integers in big endian
  #   order
  private_class_method def self.bytes_from_number(number)
    [].tap do |bytes|
      bytes.unshift(0) and next if number.zero?

      while number.positive?
        bytes.unshift(number & 0xFF)
        number >>= 8
      end
    end
  end

  # Encodes a byte array into a lexicographically sortable array of characters
  #
  # @api private
  # @private
  #
  # @param bytes [Array<Integer>] the bytes to encode, in big endian order
  # @return [Array<String>] the bytes encoded to the lexicographic base32 alphabet
  private_class_method def self.encode(bytes)
    result = Array.new((bytes.length * 8 + 4) / 5, 0)
    offset = 0

    while bytes.length.positive?
      length = bytes.length
      buffer =
        case length
        when 1, 2 then Array.new(bytes.length * 2, 0)
        when 3, 4 then Array.new(bytes.length * 2 - 1, 0)
        else Array.new(8, 0)
        end

      if length > 4
        buffer[7] = bytes[4] & 0x1F
        buffer[6] = bytes[4] >> 5
      end

      if length >= 4
        buffer[6] |= (bytes[3] << 3) & 0x1F
        buffer[5] = (bytes[3] >> 2) & 0x1F
        buffer[4] = bytes[3] >> 7
      end

      if length >= 3
        buffer[4] |= (bytes[2] << 1) & 0x1F
        buffer[3] = (bytes[2] >> 4) & 0x1F
      end

      if length >= 2
        buffer[3] |= (bytes[1] << 4) & 0x1F
        buffer[2] = (bytes[1] >> 1) & 0x1F
        buffer[1] = (bytes[1] >> 6) & 0x1F
      end

      if length >= 1
        buffer[1] |= (bytes[0] << 2) & 0x1F
        buffer[0] = bytes[0] >> 3
      end

      buffer.each_with_index do |byte, i|
        result[i + offset] = LEXICOGRAPHIC_BASE32[byte & 31]
      end

      break if bytes.length < 5

      bytes = bytes[5..]
      offset += 8
    end

    result
  end

  # Validates and coerces an argument for suitability for conversion
  #
  # @api private
  # @private
  #
  # @param object [Object] the object to validate and coerce
  # @return [Integer] the object converted into an integer timestamp
  # @raise [ArgumentError] when the object is unsuitable for conversion
  private_class_method def self.maybe_coerce_time(object)
    if object.is_a?(Integer) || object.is_a?(Float)
      raise ArgumentError, "#{object} must be non-negative" if object.negative?
      object.to_i
    elsif object.is_a?(Time)
      raise ArgumentError, "#{object.iso8601} must be after #{EPOCH.iso8601}" unless EPOCH <= object
      object.to_i
    elsif object.respond_to?(:to_time)
      maybe_coerce_time object.to_time
    else
      raise(
        ArgumentError,
        "#{object.inspect} must be one of Integer, Float, Time, or respond to `#to_time'"
      )
    end
  end
end

require_relative "core_ext/kernel"
